core/arch/arm/plat-imx/pm/psci-suspend-imx7.S

```diff
@@ -0,0 +1,733 @@
+	str	r7, [r11, #MX7D_GPC_IMR4]
+
+	/*
+	 * now delay for a short while (3usec)
+	 * ARM is at 1GHz at this point
+	 * so a short loop should be enough.
+	 * this delay is required to ensure that
+	 * the RBC counter can start counting in
+	 * case an interrupt is already pending
+	 * or in case an interrupt arrives just
+	 * as ARM is about to assert DSM_request.
+	 */
+	ldr	r7, =2000
+rbc_loop:
+	subs	r7, r7, #0x1
+	bne	rbc_loop
+wfi:
+	/* Enter stop mode */
+	wfi
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+	nop
+
+	mov	r5, #0x0
+
+	ldr	r11, [r0, #PM_INFO_MX7_GPC_V_OFF]
+	ldr	r7, [r11, #GPC_PGC_FM]
+	cmp	r7, #0
+	beq	wfi_ddr_self_refresh_out
+
+	ddr_exit_retention
+	b	wfi_ddr_retention_out
+wfi_ddr_self_refresh_out:
+	ddrc_exit_self_refresh
+wfi_ddr_retention_out:
+
+	/* check whether it is a standby mode */
+	ldr	r11, [r0, #PM_INFO_MX7_GPC_V_OFF]
+	ldr	r7, [r11, #GPC_PGC_C0]
+	cmp	r7, #0
+	beq	standby_out
+
+	ldr	r11, [r0, #PM_INFO_MX7_GIC_DIST_V_OFF]
+	ldr	r7, =0x1
+	ldr	r8, =0x1000
+	str	r7, [r11, r8]
+
+	restore_ttbr
+standby_out:
+	pop	{r4-r12}
+	/* return to suspend finish */
+	bx	lr
+
+resume:
+	write_iciallu
+	write_bpiall
+	mov     r6, #(SCTLR_I | SCTLR_Z)
+	write_sctlr r6
+	isb
+
+	/* switch to monitor mode from svc mode */
+	mov	r3, #CPSR_MODE_MON
+	mov	r4, #(BIT32(6) | BIT32(7))
+	orr	r3, r3, r4
+	msr	cpsr, r3
+	nop
+	nop
+	nop
+
+	/* get physical resume address from pm_info. */
+	ldr	lr, [r0, #PM_INFO_MX7_RESUME_ADDR_OFF]
+	/* clear core0's entry and parameter */
+	ldr	r11, [r0, #PM_INFO_MX7_SRC_P_OFF]
+	mov	r7, #0x0
+	str	r7, [r11, #MX7_SRC_GPR1]
+	str	r7, [r11, #MX7_SRC_GPR2]
+
+	mov	r5, #0x1
+
+	ldr	r11, [r0, #PM_INFO_MX7_GPC_P_OFF]
+	ldr	r7, [r11, #GPC_PGC_FM]
+	cmp	r7, #0
+	beq	dsm_ddr_self_refresh_out
+
+	ddr_exit_retention
+	b	dsm_ddr_retention_out
+dsm_ddr_self_refresh_out:
+	ddrc_exit_self_refresh
+dsm_ddr_retention_out:
+
+	bx	lr
+UNWIND( .fnend)
+END_FUNC imx7_suspend
+
+FUNC ca7_cpu_resume, :
+UNWIND( .fnstart)
+	mov     r0, #0                  @ ; write the cache size selection register to be
+	write_csselr r0			@ ; sure we address the data cache
+	isb                             @ ; isb to sync the change to the cachesizeid reg
+
+_inv_dcache_off:
+	mov     r0, #0                  @ ; set way number to 0
+_inv_nextway:
+	mov     r1, #0                  @ ; set line number (=index) to 0
+_inv_nextline:
+	orr     r2, r0, r1                          @ ; construct way/index value
+	write_dcisw r2				    @ ; dcisw invalidate data or unified cache line by set/way
+	add     r1, r1, #1 << LINE_FIELD_OFFSET     @ ; increment the index
+	cmp     r1, #1 << LINE_FIELD_OVERFLOW       @ ; look for overflow out of set field
+	bne     _inv_nextline
+	add     r0, r0, #1 << WAY_FIELD_OFFSET      @ ; increment the way number
+	cmp     r0, #0                              @ ; look for overflow out of way field
+	bne     _inv_nextway
+
+	dsb                             @ ; synchronise
+	nop
```

Comment:
Wouldn't an `isb` instead of these two `nop` be more appropriate?

Reply:
yes.